src/ClientVariables.cpp

```diff
@@ -0,0 +1,934 @@
+        QString value = jsonObj.value("value").toString();
+
+        if (key.isEmpty()) {
+            qWarning() << "ClientVariables::sendClientVariablesUpdate: Missing or invalid 'name' field";
+            continue;
+        }
+
+        if (source != ClientVariables::SourceRequest && !clientVariablesRequested.contains(key)) {
+            qDebug() << "We did not update client variable" << key << "because the server did not request it yet";
+            continue;
+        }
+
+        bool available = !mpHost->mEnableMNES; // Default availability state
+
+        if (protectedVariables.contains(key)) {
+            const auto &[updatable, behaviour, translation] = protectedVariables[key];
+
+            if (behaviour == ClientVariables::DataSharingBehaviour::Block) {
+                addResponse(key, false, false);
+            } else {
+                available = (behaviour == ClientVariables::DataSharingBehaviour::Share);
+
+                if (!available && source == ClientVariables::SourceRequest) {
+                    requested.insert(translation, purpose);
+                }
+
+                if (clientVariablesDataMap.contains(key)) {
+                    addResponse(key, available, updatable, !available, value);
+                } else {
+                    addResponse(key, available, updatable, !available, value, purpose);
+                }
+            }
+        } else if (nonProtectedVariables.contains(key)) {
+            if (clientVariablesDataMap.contains(key)) {
+                addResponse(key, true, nonProtectedVariables[key], false, value);
+            } else {
+                addResponse(key, true, nonProtectedVariables[key]);
+            }
+        } else {
+            addResponse(key, false, false);
+        }
+    }
+
+    if (source == ClientVariables::SourceClient && response.isEmpty()) {
+        return;
+    }
+
+    QJsonDocument doc2(response);
+    QString gmcpMessage = doc2.toJson(QJsonDocument::Compact);
+
+    std::string output;
+    output += TN_IAC;
+    output += TN_SB;
+    output += OPT_GMCP;
+    output += "Client.Variables.Update ";
+    output += mpHost->mTelnet.encodeAndCookBytes(gmcpMessage.toStdString());
+    output += TN_IAC;
+    output += TN_SE;
+
+    // Send response to server
+    mpHost->mTelnet.socketOutRaw(output);
+
+    // Post message about client variable settings
+    if (source == ClientVariables::SourceRequest && !requested.isEmpty()) {
+        const QString text = tr("\n          --> Control sharing preferences by clicking here for the Data tab in Settings <--\n\n");
+        QStringList commandList;
+        QStringList hintList;
+        bool useCurrentFormat = true;
+
+        commandList << "showSettingsTab(\"tab_data\")";
+        hintList << tr("Open the Data tab of the Settings menu");
+
+        const QString action1 = tr("[ ACTION ]- To enhance your gameplay experience, the server or script");
+        mpHost->mTelnet.postMessage(action1);
```

Comment:
:rotating_light: `cTelnet::postMessage(const QString&)` is designed to take a *multi-line* text message which is presented as a hanging paragraph with all but the first line left-aligned to the `-` that separates the starting of the message text with the `[ type ] -` at the start. Assemble the text as a single message with `\n` at the end of all but the last line - and work to a 80 character line length - including indent for the continuation lines.

**Apart from anything else, splitting the message text into separate strings makes it impossible for translators to work with them as they'd only get fragments of sentences.**

Reply:
I did try a version of this with a QStringList and one postMessage, but the colors I wanted with `[ACTION]` would not maintain over multiple lines and it would not wrap correctly for presentation purposes. Multiple `[ACTION]` are also not great for screen readers. I will take another look.

Reply:
Suggestion inspired the end result.